File: app/src/main/java/xyz/b515/schedule/ui/adapter/SpacetimeAdapter.java
package xyz.b515.schedule.ui.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import java.text.DateFormatSymbols;
import java.util.List;

import xyz.b515.schedule.R;
import xyz.b515.schedule.databinding.ItemSpacetimeBinding;
import xyz.b515.schedule.db.CourseManager;
import xyz.b515.schedule.entity.Spacetime;

/**
 * Created by ZeroGo on 2017/5/25.
 */

public class SpacetimeAdapter extends RecyclerView.Adapter<SpacetimeAdapter.SpacetimeViewHolder> {
    public List<Spacetime> items;
    private Context context;
    private CourseManager manager;
    private Boolean flag;

    public SpacetimeAdapter(List<Spacetime> items, Context context, CourseManager manager, Boolean flag) {
        this.items = items;
        this.context = context;
        this.manager = manager;
        this.flag = flag;
    }

    @Override
    public SpacetimeViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        ItemSpacetimeBinding binding = DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.item_spacetime, viewGroup, false);

        return new SpacetimeViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(SpacetimeViewHolder holder, int position) {
        holder.binding.setSpacetime(items.get(position));
        String[] weeks = DateFormatSymbols.getInstance().getShortWeekdays();
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(context, android.R.layout.simple_spinner_dropdown_item, weeks);
        holder.binding.spinner.setAdapter(adapter);
        if(!flag) {
            holder.binding.spinner.setSelection(items.get(position).getWeekday());
        }
        holder.binding.spacetimeDelete.setOnClickListener(v -> {
            new AlertDialog.Builder(context)
                    .setTitle(R.string.alert_title)
                    .setMessage(R.string.alert_message)
                    .setPositiveButton(R.string.alert_pos, (dialog, which) -> {
                        manager.deleteSpacetime(items.get(position));
                        items.remove(position);
                        notifyDataSetChanged();
                    })
                    .setNegativeButton(R.string.alert_neg, (dialog, which) -> {
                    })
                    .show();
        });
        holder.binding.spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int pos, long id) {
                items.get(position).setWeekday(pos);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
    }

    @Override
    public int getItemCount() {
        if (items == null) {
            return 0;
        }
        return items.size();

    }

    class SpacetimeViewHolder extends RecyclerView.ViewHolder {
        ItemSpacetimeBinding binding;

        SpacetimeViewHolder(View view) {
            super(view);
            binding = DataBindingUtil.getBinding(view);
        }
    }
}

File: app/src/main/kotlin/xyz/b515/schedule/ui/adapter/SpacetimeAdapter.kt
package xyz.b515.schedule.ui.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import kotlinx.android.synthetic.main.item_spacetime.view.*
import org.jetbrains.anko.alert
import xyz.b515.schedule.R
import xyz.b515.schedule.db.CourseManager
import xyz.b515.schedule.entity.Spacetime
import java.text.DateFormatSymbols
import java.util.*

/**
 * Created by ZeroGo on 2017/5/25.
 */

class SpacetimeAdapter(var items: ArrayList<Spacetime>, private val context: Context, private val manager: CourseManager, private val flag: Boolean) : RecyclerView.Adapter<SpacetimeAdapter.SpacetimeViewHolder>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int) =
            SpacetimeViewHolder(LayoutInflater.from(context).inflate(R.layout.item_spacetime, viewGroup, false))

    override fun onBindViewHolder(holder: SpacetimeViewHolder, position: Int) {
        val st = items[position]
        with(holder.v) {
            spacetime_start_week.setText(st.startWeek.toString())
            spacetime_end_week.setText(st.endWeek.toString())
            spacetime_start_time.setText(st.startTime.toString())
            spacetime_end_time.setText(st.endTime.toString())
            spacetime_place.setText(st.location)
        }

        val w = object : TextWatcher {
            fun Editable.toInt(): Int = this.toString().toIntOrNull() ?: 0
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                with(st) {
                    startWeek = holder.v.spacetime_start_week.text.toInt()
                    endWeek = holder.v.spacetime_end_week.text.toInt()
                    startTime = holder.v.spacetime_start_time.text.toInt()
                    endTime = holder.v.spacetime_end_time.text.toInt()
                    location = holder.v.spacetime_place.text.toString()
                }
            }
        }
        listOf(holder.v.spacetime_start_week, holder.v.spacetime_end_week, holder.v.spacetime_start_time,
                holder.v.spacetime_end_time, holder.v.spacetime_place).forEach { it.addTextChangedListener(w) }

        val weeks = DateFormatSymbols.getInstance().shortWeekdays.toMutableList()
        weeks.removeAt(0)
        val adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, weeks)
        holder.v.spinner.adapter = adapter
        if (!flag) {
            holder.v.spinner.setSelection(items[position].weekday - 1)
        }

        holder.v.spacetime_delete.setOnClickListener {
            context.alert(R.string.alert_title) {
                messageResource = R.string.alert_message
                positiveButton(R.string.alert_pos) { _ ->
                    manager.deleteSpacetime(items[holder.adapterPosition])
                    items.removeAt(holder.adapterPosition)
                    notifyDataSetChanged()
                }
                negativeButton(R.string.alert_neg) {}
            }.show()
        }
        holder.v.spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, pos: Int, id: Long) {
                items[holder.adapterPosition].weekday = pos + 1
            }

            override fun onNothingSelected(parent: AdapterView<*>) = Unit
        }
    }

    override fun getItemCount() = items.size

    class SpacetimeViewHolder(val v: View) : RecyclerView.ViewHolder(v)
}
